File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque 
import sys

input = sys.stdin.readline

dx = [-1, 1, 0 ,0]
dy = [0, 0, -1, 1]

def bfs(x,y):
  cnt = 1
  queue = deque() 
  queue.append((x,y))
  visited[x][y] = True

  while(queue):
    x, y = queue.popleft() 
    
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]
      if(0<=nx<n and 0<=ny<n):
        if(visited[nx][ny] == False and data[nx][ny] == 1):
          visited[nx][ny] = True
          queue.append((nx,ny))
          cnt += 1

  return cnt

n = int(input())
data = []
visited = [[False]*(n + 1) for _ in range(n + 1)]
result = []

for _ in range(n):
  data.append(list(map(int, input().strip())))

for i in range(n):
  for j in range(n):
    if visited[i][j] == False and data[i][j] == 1:
      result.append(bfs(i,j))

result.sort()
length = len(result)

print(length)

for i in range(length):
  print(result[i])

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static Queue<Integer> queue = new LinkedList<>();
    static int n;
    static int[][] data;
    static boolean[][] visited;
    static LinkedList<Integer> result = new LinkedList<>();

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        data = new int[n][n];
        visited = new boolean[n][n];

        for(int i = 0; i<n; i++){
            String[] strings = br.readLine().split("");
            for(int j = 0; j<n; j++){
                data[i][j] = Integer.parseInt(strings[j]);
            }
        }

        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++){
                if(!visited[i][j] && data[i][j] == 1){
                    result.add(bfs(i,j));
                }
            }
        }

        Collections.sort(result);

        System.out.println(result.size());
        for (int a: result) {
            System.out.println(a);
        }
        br.close();
    }

    static private int bfs(int x, int y){
        int count = 1;
        queue.add(x);
        queue.add(y);

        visited[x][y] = true;

        while(!queue.isEmpty()){
            int popX = queue.poll();
            int popY = queue.poll();

            for (int i = 0; i < 4 ; i++) {
                int nx = popX + dx[i];
                int ny = popY + dy[i];

                if((0 <= nx) && (nx < n) && (0 <= ny) && (ny < n)){
                    if(!visited[nx][ny] && data[nx][ny] == 1){
                        count++;
                        queue.add(nx);
                        queue.add(ny);
                        visited[nx][ny] = true;
                    }
                }
            }
        }
        return count;
    }
}
